### PR TITLE
Feature: Attributes auto-merge

### PR DESCRIPTION
**Pull request by tiagomapmarques**
This proposed change solves the following:
 - If a parent component updates a BioElement attribute, the element receives the notification and updates it's props automatically

Notes:
 - numbers passed on attributes will be converted to numbers on props. strings will stay as is
 - `propsFromAttributes` is now only a helper for BioElement components as `attributeChangedCallback` now updates `_props`
 - added notes on all overridable functions

### CODE REVIEW

**Comment by marwesfur** [comment]
how are `propsFromAttributes` integrated into `props`? As far as I can see, nothing happens with `propsFromAttributes`.

Your changes imply, that `TProps` is always just an assembly of all possible attributes. When we were talking about the relationshop between `props` and attributes, Jurek brought the example of `props` being a "configuration object" also containing function objects for callback etc. 

While I'm still a little bit skeptical regarding this approach for passing in callbacks, we henceforth seemingly agreed that `props` can be more then just the sum of all attributes.

**Comment by tiagomapmarques** [comment]
@marwesfur 
The `propsFromAttributes` is now just a helper for any components that extend from BioElement. I'm ok with removing it if it confuses anyone. `attributeChangedCallback` is now responsible for merging attributes with props on any attribute that is observed by the `observedAttributes` function.

My changes propose 2 things:
1) a merge between the props and any observable attribute when said attribute is changed. Props is still the de-facto way of passing down complex objects and functions. What this approach does is simplify the "API" of the component for simple usage (like passing strings or indexes) and
2) a way for parents to fully control the children *if* they need to (i.e. by changing an attribute, which gets auto-merged with props and causes a child component re-render).

This means that any web-component built with BioElement just needs to read its props, and it gets all the information merged (attributes [i.e. simple variables], defaultProps [i.e. default values for attributes/props] and actual props [i.e. complex objects or functions passed from the parent]).

Basically, what I am proposing is a unified way for any component to read/use it's "input data" and a simple way for parents to pass down simple information 😄

**Comment by marwesfur** [comment]
@tiagomapmarques 

thanks for clarifying. What I overlooked is that `attributeChangedCallback` is also initially called for all attributes (not only when those change later on), so that `props`, as you say, then contains the full set of attribute values. 

I would then really suggest to remove `propsFromAttributes` or make it available as an optional mixin.

What remains, though, is my slight scepticism, that now there is no way to enforce attributes (and their value types) to really correspond to `TProps` (that was the intention with the original `propsFromAttributes`: The component author is responsible to transform the attributes to a `TProps` value in a way the TS compiler is able to check).

**Comment by tiagomapmarques** [comment]
Agreed on removing `propsFromAttributes`. It's definitely more confusing than helpful.

As for the type-checking, I love TS so I understand the drive to enforce typechecking everywhere. But when we're dealing with html attributes, TS type-enforcing always goes out the window because regardless of what you put in, it will always come out as a string. And this is not "known to TS" on compilation, i.e. if you say an html attribute is a number, TS will compile fine, but on runtime it's going to be a string.

That said, I dont understand why these changes remove the ability to enforce types on attributes. Can you please explain?

**Comment by marwesfur** [comment]
@tiagomapmarques 

let me sketch what we had in mind during our workshop (written here, no guarantee that it directly compiles):

1. One of the first steps when creating a component is to define it's "input interface":

    ~~~~
    interface SliderProps {
       selectedSlides: number[];  // made up a little bit to demonstrate my point of enforcing type constraints
    }
    ~~~~

2. At a later step the developer has to decide what attributes to support; she should then implement `propsFromAttributes` to map those attributes to `SliderProps`

    ~~~~
    get propsFromAttributes(): SliderProps {
      return {
        selectedSlides: this.hasAttribute('selectedSlides') 
          ? this.getAttribute('selectedSlides').split(',').map(parseInt) 
          : []
      };
    }
    ~~~~

    Now TS can easily check if the mapping is done right (cause it knows, that `getAttribute` returns a `string`, `split` turns it into `string[]` and `parseInt` returns a `number`.

That approach clearly had plain JS and Angular as "targets" in mind. Things are obviously done a little bit different in vue.js, react and plain html, where attributes with string values are the primary input interface;  that reduces `TProps` to be a flat collection of these attributes with only string or number values.

I'm comfortable with either way, I think we should just make this decision explicit -- so that we can also write it down in our documentation.

@jurekbarth  Can you agree?

**Comment by tiagomapmarques** [comment]
@marwesfur
I see your point now.
You can still use my changes that way, just that your second step is done automatically.

But if you still need to do some logic before putting them in props, you can always override the `attributeChangedCallback` and do the same logic there. IMO overriding this function seems more natural as it's a known web-component lifecycle function - as opposed to overriding a BioElement specific function (`propsFromAttributes`).

**Comment by jurekbarth** [comment]
@tiagomapmarques When developing a component and you've used `propsFromAttributes` they are typed because you (as a developer) know what to expect. When you now use that component you will now what the author expected as an input. With that automagic approach you don't know anymore... @tiagomapmarques you're right regarding attributes are always strings. But if the author expected a simple object or a number of a certain format, with that approach you need to read how it's used right.

**Comment by marwesfur** [comment]
I vote for merging this PR (when the now slightly misleading `getPropsFromAttribute` is removed), but we definitely need good documentation regarding attributes and props (ofc, we needed this in the first place, but this discussion has again clearly emphasized that need).

**Comment by marwesfur** [comment]
here is one more suggestion (addressing Marcs wish to have id="123" stay a string): instead of `observableAttributes`, lets have a `biotopeAttributes`, where each attribute can also be supplied an (optional) converter func, that is then applied to the attribute value before merging into props:

~~~~
static bioElements = [ 
  'aSimpleAttrtibute, 
  {name: 'aCommaSeparatedAttribute', convert: _ => _.split(',')},  // needs a more robust impl. for prod
  {name: 'id', convert: identity}
];

~~~~

**Comment by tiagomapmarques** [comment]
@marwesfur @jurekbarth @SheepFromHeaven
PTAL

File: index.ts [hunk 1]
```diff
@@ -1,35 +1,58 @@
...
   };
 
-  attributeChangedCallback() {
-    this.onPropsChanged();
+  attributeChangedCallback(name: string, _: string, newValue: string): void {
+    const attribute = BioElement.bioAttributes.find(attr => attributeName(attr) === newValue);
+    this.props = {
+      ...(this.props as any),
+      [name]: typeof attribute === 'string'
+        ? parseAttribute(newValue)
```
Comment (marwesfur) [inline index.ts]: if I remember correctly this is not what @SheepFromHeaven wanted: if the attribute does not have a custom convert function, leave it as a string.

Comment (tiagomapmarques) [reply index.ts]: From what I understood, if the developer says nothing about the attribute (i.e., if it's just a string), then we parse it. If not, we're expecting a converter function. Right?

Comment (marwesfur) [reply index.ts]: @SheepFromHeaven could you comment on that? You had that "an ID should stay a string even if it is just digits"-example.

**Comment by marwesfur** [comment]
@SheepFromHeaven @jurekbarth could we finally merge this, please? I want to build on top of it. Thanks!

File: index.ts (hunk 1 above)
Comment (jurekbarth) [reply index.ts]: String by default, anything else custom function 👍

Comment (tiagomapmarques) [reply index.ts]: Reverted 👍
